src/generic/stage2/numberparsing.h

```diff
@@ -233,109 +244,131 @@ really_inline bool is_made_of_eight_digits_fast(const char *chars) {
-      found_integer(i, src);
-#endif
-      writer.append_s64(i);
-    } else {
-#ifdef JSON_TEST_NUMBERS // for unit testing
-      found_unsigned_integer(i, src);
-#endif
-      writer.append_u64(i);
-    }
+
+  // e[+-] must be followed by a number
+  if (!is_integer(*p)) { return INVALID_NUMBER(src); }
+  unsigned char digit = static_cast<unsigned char>(*p - '0');
+  int64_t exp_number = digit;
+  p++;
+  if (is_integer(*p)) {
+    digit = static_cast<unsigned char>(*p - '0');
+    exp_number = 10 * exp_number + digit;
+    ++p;
   }
-  return is_structural_or_whitespace(*p);
+  if (is_integer(*p)) {
+    digit = static_cast<unsigned char>(*p - '0');
+    exp_number = 10 * exp_number + digit;
+    ++p;
+  }
+  while (is_integer(*p)) {
+    // we need to check for overflows; we refuse to parse this
+    if (exp_number > 0x100000000) { return INVALID_NUMBER(src); }
+    digit = static_cast<unsigned char>(*p - '0');
+    exp_number = 10 * exp_number + digit;
+    ++p;
+  }
+  exponent += (neg_exp ? -exp_number : exp_number);
+  return true;
 }
 
 template<typename W>
-bool slow_float_parsing(UNUSED const char * src, W writer) {
-  double d;
-  if (parse_float_strtod(src, &d)) {
-    writer.append_double(d);
-#ifdef JSON_TEST_NUMBERS // for unit testing
-    found_float(d, (const uint8_t *)src);
-#endif
-    return true;
+really_inline bool write_float(const uint8_t *const src, bool negative, uint64_t i, const char * start_digits, int digit_count, int64_t exponent, W &writer) {
+  // If we frequently had to deal with long strings of digits,
+  // we could extend our code by using a 128-bit integer instead
+  // of a 64-bit integer. However, this is uncommon in practice.
+  // digit count is off by 1 because of the decimal (assuming there was one).
+  if (unlikely((digit_count-1 >= 19))) { // this is uncommon
+    // It is possible that the integer had an overflow.
+    // We have to handle the case where we have 0.0000somenumber.
+    const char *start = start_digits;
+    while ((*start == '0') || (*start == '.')) {
+      start++;
+    }
+    // we over-decrement by one when there is a '.'
+    digit_count -= int(start - start_digits);
+    if (digit_count >= 19) {
+      // Ok, chances are good that we had an overflow!
+      // this is almost never going to get called!!!
+      // we start anew, going slowly!!!
+      // This will happen in the following examples:
+      // 10000000000000000000000000000000000000000000e+308
+      // 3.1415926535897932384626433832795028841971693993751
+      //
+      bool success = slow_float_parsing((const char *) src, writer);
+      // The number was already written, but we made a copy of the writer
+      // when we passed it to the parse_large_integer() function, so
+      writer.skip_double();
+      return success;
+    }
   }
-#ifdef JSON_TEST_NUMBERS // for unit testing
-  found_invalid_number((const uint8_t *)src);
-#endif
-  return false;
+  // TODO unlikely wraps the wrong thing here
```

Comment:
Same question... Given that you have identified that the unlikely should surround the whole condition, and not just part of it, why not fix it up?

Reply:
This one, I just forgot to check. I have learned that even little refactors like this need to be made and evaluated separately. My mental model of the compiler *still* doesn't match what happens in a lot of cases; I'm no longer sure there is a reasonably-sized mental model that *can* :)

Reply:
WTF? It's slower when you do this? I am so confused. I'mma rerun the numbers, but here's what I got:

| File                                | master Cycles | PR Cycles | +Throughput | master Instr. | PR Instr. | -Instr. | master Branch Misses | PR Branch Misses | -Branch Misses | master Cache Misses | PR Cache Misses | -Cache Misses | master Cache Refs | PR Cache Refs | -Cache Refs |
| ---                                 |   --: |   --: |   --: |   --: |   --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |
| repeat                              |  50.9 |  50.8 |    0% | 187.8 | 187.8 |    0% |      1 |      1 |    0% |      0 |      0 |    0% |     25 |     30 |  -20% |
| twitter_api_compact_response        |  41.5 |  41.4 |    0% | 153.1 | 153.1 |    0% |      1 |      1 |    0% |      0 |      0 |    0% |      3 |      4 |  -33% |
| tree-pretty                         |  37.0 |  36.9 |    0% | 140.6 | 140.6 |    0% |      6 |      1 |   83% |      0 |      0 |    0% |    236 |    463 |  -96% |
| github_events                       |  34.3 |  34.3 |    0% | 129.8 | 129.8 |    0% |      4 |      4 |    0% |      0 |      0 |    0% |   1849 |   1882 |   -1% |
| google_maps_api_compact_response    |  50.1 |  50.1 |    0% | 190.3 | 190.3 |    0% |      1 |      1 |    0% |      6 |      0 |    0% |     63 |     46 |   26% |
| twitter_api_response                |  37.6 |  37.6 |    0% | 140.1 | 140.1 |    0% |      1 |      1 |    0% |      0 |      0 |    0% |     34 |      9 |   73% |
| citm_catalog                        |  34.1 |  34.1 |    0% | 130.7 | 130.7 |    0% |    164 |    164 |    0% |      2 |      0 |    0% |  76725 |  76646 |    0% |
| twitter_timeline                    |  40.6 |  40.6 |    0% | 153.5 | 153.5 |    0% |      4 |      3 |   25% |      0 |      0 |    0% |    326 |    619 |  -89% |
| mesh.pretty                         |  35.7 |  35.8 |    0% | 137.7 | 137.7 |    0% |     39 |     37 |    5% |      0 |      0 |    0% |  74481 |  74302 |    0% |
| random                              |  57.6 |  57.7 |    0% | 209.3 | 209.3 |    0% |   1013 |    980 |    3% |      0 |      0 |    0% |  29778 |  29768 |    0% |
| numbers                             |  38.8 |  38.9 |    0% | 144.3 | 144.3 |    0% |    125 |    132 |   -5% |      0 |      0 |    0% |   7197 |   7456 |   -3% |
| gsoc-2018                           |  35.7 |  35.9 |    0% | 101.1 | 101.1 |    0% |  14683 |  15036 |   -2% |  18225 |  24015 |  -31% | 122163 | 122212 |    0% |
| google_maps_api_response            |  36.1 |  36.4 |    0% | 137.2 | 137.2 |    0% |      1 |      3 | -200% |      0 |      0 |    0% |     24 |    355 | -1379% |
| apache_builds                       |  36.7 |  37.0 |    0% | 139.3 | 139.3 |    0% |     21 |     33 |  -57% |      0 |      0 |    0% |   5757 |   5510 |    4% |
| canada                              |  45.0 |  45.4 |    0% | 157.6 | 157.6 |    0% |   5518 |   5301 |    3% |   7680 |  10664 |  -38% | 122181 | 122279 |    0% |
| mesh                                |  45.9 |  46.3 |    0% | 174.2 | 174.2 |    0% |    725 |    821 |  -13% |      0 |      0 |    0% |  45939 |  46015 |    0% |
| marine_ik                           |  53.8 |  54.7 |   -1% | 180.9 | 180.9 |    0% |   8882 |   8772 |    1% |  72327 |  79961 |  -10% | 191265 | 191284 |    0% |
| twitter                             |  43.1 |  43.8 |   -1% | 148.8 | 148.8 |    0% |   1721 |   1875 |   -8% |      0 |      0 |    0% |  29042 |  29139 |    0% |
| instruments                         |  37.8 |  38.5 |   -1% | 143.1 | 143.1 |    0% |     95 |    184 |  -93% |      0 |      0 |    0% |  11239 |  11147 |    0% |
| update-center                       |  43.4 |  45.0 |   -3% | 142.1 | 142.1 |    0% |   2541 |   2895 |  -13% |      0 |      0 |    0% |  26901 |  27028 |    0% |
| twitterescaped                      |  41.1 |  42.9 |   -4% | 138.1 | 138.1 |    0% |   1882 |   2307 |  -22% |      0 |      0 |    0% |  26811 |  26851 |    0% |

Reply:
It's consistent over multiple runs. Same files get worse by roughly the same amounts, so it's not noise. Removing the unlikely altogether makes things even worse:

| File                                | Blocks | master Cycles | PR Cycles | +Throughput | master Instr. | PR Instr. | -Instr. | master Branch Misses | PR Branch Misses | -Branch Misses | master Cache Misses | PR Cache Misses | -Cache Misses | master Cache Refs | PR Cache Refs | -Cache Refs |
| ---                                 |    --- |   --: |   --: |   --: |   --: |   --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |
| github_events                       |   1017 |  34.3 |  34.3 |    0% | 129.8 | 129.8 |    0% |      4 |      7 |  -75% |      0 |      0 |    0% |   1849 |   1543 |   16% |
| citm_catalog                        |  26987 |  34.1 |  34.1 |    0% | 130.7 | 130.7 |    0% |    164 |    176 |   -7% |      2 |    162 | -8000% |  76725 |  71639 |    6% |
| canada                              |  35172 |  45.0 |  45.1 |    0% | 157.6 | 157.6 |    0% |   5518 |   5323 |    3% |   7680 |  10140 |  -32% | 122181 | 115262 |    5% |
| mesh.pretty                         |  24646 |  35.7 |  35.8 |    0% | 137.7 | 137.7 |    0% |     39 |     41 |   -5% |      0 |      1 |    0% |  74481 |  69339 |    6% |
| apache_builds                       |   1988 |  36.7 |  36.8 |    0% | 139.3 | 139.3 |    0% |     21 |     20 |    4% |      0 |      0 |    0% |   5757 |   5021 |   12% |
| mesh                                |  11306 |  45.9 |  46.4 |   -1% | 174.2 | 174.2 |    0% |    725 |    883 |  -21% |      0 |      0 |    0% |  45939 |  42984 |    6% |
| marine_ik                           |  46616 |  53.8 |  54.5 |   -1% | 180.9 | 180.9 |    0% |   8882 |   8704 |    2% |  72327 |  78559 |   -8% | 191265 | 186405 |    2% |
| gsoc-2018                           |  51997 |  35.7 |  36.2 |   -1% | 101.1 | 101.1 |    0% |  14683 |  14898 |   -1% |  18225 |  19603 |   -7% | 122163 | 116354 |    4% |
| instruments                         |   3442 |  37.8 |  38.6 |   -2% | 143.1 | 143.1 |    0% |     95 |    218 | -129% |      0 |      0 |    0% |  11239 |  10142 |    9% |
| twitter                             |   9867 |  43.1 |  44.2 |   -2% | 148.8 | 148.8 |    0% |   1721 |   1855 |   -7% |      0 |      0 |    0% |  29042 |  27228 |    6% |
| random                              |   7976 |  57.6 |  59.1 |   -2% | 209.3 | 209.3 |    0% |   1013 |   1019 |    0% |      0 |      0 |    0% |  29778 |  27842 |    6% |
| numbers                             |   2345 |  38.8 |  40.1 |   -3% | 144.3 | 144.3 |    0% |    125 |    260 | -108% |      0 |      0 |    0% |   7197 |   6968 |    3% |
| update-center                       |   8330 |  43.4 |  45.0 |   -3% | 142.1 | 142.1 |    0% |   2541 |   2881 |  -13% |      0 |      0 |    0% |  26901 |  25266 |    6% |
| twitterescaped                      |   8787 |  41.1 |  42.7 |   -3% | 138.1 | 138.1 |    0% |   1882 |   2248 |  -19% |      0 |      0 |    0% |  26811 |  25114 |    6% |

Reply:
I can't think of another experiment to do here to explain this result. I'm going to leave the TODO in with a comment about this fact, and be done. :(

Reply:
Here's what happens when you remove the redundant check:

| File                                | Blocks | master Cycles | PR Cycles | +Throughput | master Instr. | PR Instr. | -Instr. | master Branch Misses | PR Branch Misses | -Branch Misses | master Cache Misses | PR Cache Misses | -Cache Misses | master Cache Refs | PR Cache Refs | -Cache Refs |
| ---                                 |    --- |   --: |   --: |   --: |   --: |   --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |    --: |    --: |   --: |
| citm_catalog                        |  26987 |  34.1 |  34.1 |    0% | 130.7 | 130.7 |    0% |    164 |    179 |   -9% |      2 |      7 | -250% |  76725 |  71783 |    6% |
| github_events                       |   1017 |  34.3 |  34.4 |    0% | 129.8 | 129.8 |    0% |      4 |      4 |    0% |      0 |      0 |    0% |   1849 |   1164 |   37% |
| mesh.pretty                         |  24646 |  35.7 |  35.8 |    0% | 137.7 | 137.7 |    0% |     39 |     42 |   -7% |      0 |      0 |    0% |  74481 |  69348 |    6% |
| random                              |   7976 |  57.6 |  57.8 |    0% | 209.3 | 209.3 |    0% |   1013 |   1005 |    0% |      0 |      0 |    0% |  29778 |  27763 |    6% |
| marine_ik                           |  46616 |  53.8 |  54.1 |    0% | 180.9 | 180.9 |    0% |   8882 |   8771 |    1% |  72327 |  74782 |   -3% | 191265 | 186677 |    2% |
| canada                              |  35172 |  45.0 |  45.3 |    0% | 157.6 | 157.6 |    0% |   5518 |   5354 |    2% |   7680 |  10541 |  -37% | 122181 | 115501 |    5% |
| apache_builds                       |   1988 |  36.7 |  37.0 |    0% | 139.3 | 139.3 |    0% |     21 |     31 |  -47% |      0 |      0 |    0% |   5757 |   5118 |   11% |
| twitter                             |   9867 |  43.1 |  43.5 |    0% | 148.8 | 148.8 |    0% |   1721 |   1752 |   -1% |      0 |      0 |    0% |  29042 |  27124 |    6% |
| mesh                                |  11306 |  45.9 |  46.4 |   -1% | 174.2 | 174.2 |    0% |    725 |    884 |  -21% |      0 |      0 |    0% |  45939 |  43025 |    6% |
| gsoc-2018                           |  51997 |  35.7 |  36.4 |   -2% | 101.1 | 101.1 |    0% |  14683 |  14877 |   -1% |  18225 |  18775 |   -3% | 122163 | 116368 |    4% |
| update-center                       |   8330 |  43.4 |  44.9 |   -3% | 142.1 | 142.1 |    0% |   2541 |   2810 |  -10% |      0 |      0 |    0% |  26901 |  25241 |    6% |
| instruments                         |   3442 |  37.8 |  39.2 |   -3% | 143.1 | 143.1 |    0% |     95 |    300 | -215% |      0 |      0 |    0% |  11239 |  10321 |    8% |
| numbers                             |   2345 |  38.8 |  40.3 |   -4% | 144.3 | 144.3 |    0% |    125 |    275 | -120% |      0 |      0 |    0% |   7197 |   6988 |    2% |
| twitterescaped                      |   8787 |  41.1 |  42.9 |   -4% | 138.1 | 138.1 |    0% |   1882 |   2293 |  -21% |      0 |      0 |    0% |  26811 |  25046 |    6% |

Reply:
On the theory that the redundant check gives the compiler information it didn't otherwise have, I also replaced the if() with SIMDJSON_ASSUME(is_integer(*p)). (The second if, not the one marked redundant here--it's not actually the redundant one.) That causes us to do `if (!is_integer(*p)) { __builtin_unreachable(); }`, which will tell the compiler that is_integer(*p) would succeed, but removes the actual call. The result is exactly the same as when you remove the if(): things got slower.

Reply:
So, for now I'm going to leave a big honking comment. Again.